object/slicer/slicer.go

```diff
@@ -0,0 +1,596 @@
+type ObjectWriter interface {
+	// InitDataStream initializes and returns a stream of writable data associated
+	// with the object according to its header. Provided header includes at least
+	// container, owner and object ID fields.
+	InitDataStream(header object.Object) (dataStream io.Writer, err error)
+}
+
+// Slicer converts input raw data streams into NeoFS objects. Working Slicer
+// must be constructed via New.
+type Slicer struct {
+	signer neofscrypto.Signer
+
+	cnr cid.ID
+
+	owner user.ID
+
+	w ObjectWriter
+
+	opts Options
+
+	sessionToken *session.Object
+}
+
+// New constructs Slicer which writes sliced ready-to-go objects owned by
+// particular user into the specified container using provided ObjectWriter.
+// All objects are signed using provided neofscrypto.Signer.
+//
+// If ObjectWriter returns data streams which provide io.Closer, they are closed
+// in Slicer.Slice after the payload of any object has been written. In this
+// case, Slicer.Slice fails immediately on Close error.
+//
+// Options parameter allows you to provide optional parameters which tune
+// the default Slicer behavior. They are detailed below.
+//
+// If payload size limit is specified via Options.SetObjectPayloadLimit,
+// outgoing objects has payload not bigger than the limit. NeoFS stores the
+// corresponding value in the network configuration. Ignore this option if you
+// don't (want to) have access to it. By default, single object is limited by
+// 1MB. Slicer uses this value to enforce the maximum object payload size limit
+// described in the NeoFS Specification. If the total amount of data exceeds the
+// specified limit, Slicer applies the slicing algorithm described within the
+// same specification. The outcome will be a group of "small" objects containing
+// a chunk of data, as well as an auxiliary linking object. All derived objects
+// are written to the parameterized ObjectWriter. If the amount of data is
+// within the limit, one object is produced. Note that Slicer can write multiple
+// objects, but returns the root object ID only.
+//
+// If current NeoFS epoch is specified via Options.SetCurrentNeoFSEpoch, it is
+// written to the metadata of all resulting objects as a creation epoch.
+//
+// See also NewSession.
+func New(signer neofscrypto.Signer, cnr cid.ID, owner user.ID, w ObjectWriter, opts Options) *Slicer {
+	return &Slicer{
+		signer: signer,
+		cnr:    cnr,
+		owner:  owner,
+		w:      w,
+		opts:   opts,
+	}
+}
+
+// NewSession creates Slicer which generates objects within provided session.
+// NewSession work similar to New with the detail that the session issuer owns
+// the produced objects. Specified session token is written to the metadata of
+// all resulting objects. In this case, the object is considered to be created
+// by a proxy on behalf of the session issuer.
+func NewSession(signer neofscrypto.Signer, cnr cid.ID, token session.Object, w ObjectWriter, opts Options) *Slicer {
```

Comment:
How about removing `opts` (and `Options`) and adding a `limit` (for `objectPayloadLimit`) here (which can be 0 for default one)...

Reply:
dislike, structured options are extendable and easier for defaults: code with explicit `0`, `""`, `nil`, ... args looks much less clear

Reply:
My rationale is rather simple, a session can't be created without some value for `limit`. And this value is inherently the same during the whole session lifetime.

And yes, functional options is not my favorite thing. I can live with them, but I'd rather avoid them for simple things.

Reply:
Both are optional, leaving as is.